[Feature] Support Metadata filtering with Azure AI Search

fb33 opened the pull request:
<!-- Thank you so much for your contribution! -->
<!-- Please fill in all the sections below. -->

<!-- Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples. -->
<!-- Please note that PRs with breaking changes will be rejected. -->
<!-- Please note that PRs without tests will be rejected. -->

<!-- Please note that PRs will be reviewed based on the priority of the issues they address. -->
<!-- We ask for your patience. We are doing our best to review your PR as quickly as possible. -->
<!-- Please refrain from pinging and asking when it will be reviewed. Thank you for understanding! -->


## Issue
https://github.com/langchain4j/langchain4j/issues/1263


## Change
add a mapper of filter to azure ai search filter string base on the default index structure with metadata.



## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

fb33 committed:
feat(filter): create a filter mapper

Create a AzureAiSearchFilterMapper to manage azure filter query format.

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 committed:
feat(filter): update retrieve to be able to use filter

Add a filter in the constructor.
Switch from the deprecated method findRelevant to the new search method.
EmbeddingSearchRequest allows to pass the filter to use.

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented:
Ping @jdubois

fb33 committed:
feat(filter): improve mapper declaration

create an interface to open the mapping to other structure of index.
change static declaration to instance, and manage mapper with builders

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented:
@jdubois  I've made a few changes to make it possible to create and use specific mapper implementations.
I change AzureAiSearchContentRetrieverIT to extend EmbeddingStoreWithFilteringIT as mention by @langchain4j but I don't run it 😁

fb33 commented:
I don't have the well configured environment for that... The IT in azure-ai-search module are not managed with TestContainer so I need to have the good env in my Azure subscription.

jdubois commented:
@fb33 I've run it, unfortunately I'm getting errors, here's a screenshot
![Screenshot 2024-06-18 at 17 21 00](https://github.com/langchain4j/langchain4j/assets/316835/7a7b58c8-9747-4642-ae05-4fc114cf63f3)

fb33 commented:
~~Ok, I'll have a look. Do I need some specific index definition ?
I quickly ran it this morning and all the tests was failed. I didn't have time to look for the root cause.~~
I create an empty azure ai search resource and the IT from the main branch are Ok, so I'll be able to investigate.

fb33 committed:
fix(filter): fix notEqualTo and notIsIn

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented:
@jdubois It should be ok now.

jdubois commented:
@fb33 I'm still getting 2 errors:

DefaultAzureAiSearchFilterMapperTest.map_handlesComplexFilter()

```
org.opentest4j.AssertionFailedError: 
Expected :(metadata/attributes/any(k: k/key eq 'key1' and k/value eq 'value1') and (metadata/attributes/any(k: k/key eq 'key2' and not search.in(k/value, ('value2, value3'))) or metadata/attributes/any(k: k/key eq 'key3' and k/value gt '100')))
Actual   :(metadata/attributes/any(k: k/key eq 'key1' and k/value eq 'value1') and ((not metadata/attributes/any(k: k/key eq 'key2' and search.in(k/value, ('value2, value3')))) or metadata/attributes/any(k: k/key eq 'key3' and k/value gt '100')))
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1141)
	at dev.langchain4j.rag.content.retriever.azure.search.DefaultAzureAiSearchFilterMapperTest.map_handlesComplexFilter(DefaultAzureAiSearchFilterMapperTest.java:75)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
```

And DefaultAzureAiSearchFilterMapperTest.map_handlesIsNotIn() : 
```
org.opentest4j.AssertionFailedError: 
Expected :metadata/attributes/any(k: k/key eq 'key1' and not search.in(k/value, ('value1, value2')))
Actual   :(not metadata/attributes/any(k: k/key eq 'key1' and search.in(k/value, ('value1, value2'))))
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1141)
	at dev.langchain4j.rag.content.retriever.azure.search.DefaultAzureAiSearchFilterMapperTest.map_handlesIsNotIn(DefaultAzureAiSearchFilterMapperTest.java:63)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
```

fb33 committed:
fix(filter): align TU with the new implementation

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented:
Sorry, I was focus on the IT test, I forgot to align my TU... It's done now.

fb33 commented:
@langchain4j @jdubois, did this PR go to the right way for you ?

jdubois commented:
@fb33 yes, I confirm all the tests pass on my side!

fb33 commented on langchain4j-azure-ai-search/src/main/java/dev/langchain4j/rag/content/retriever/azure/search/AzureAiSearchFilterMapper.java:
```diff
@@ -0,0 +1,7 @@
+
+import dev.langchain4j.store.embedding.filter.Filter;
+
+public interface AzureAiSearchFilterMapper {
```
My idea is that the default mapper I wrote is based on the default structure of index created in azure ai search. I mean there is an attribute named "metadata" with an attribute named "attributes" that represent a map key/value.
But if someone define an other structure and want to filter then he'll have to adapt the format of filter string generated.

fb33 commented on langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java:
```diff
@@ -256,29 +263,34 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
-            float[] embeddingArray = doublesListToFloatArray(embeddingList);
-            Embedding embedding = Embedding.from(embeddingArray);
+            Embedding embedding = null;
+            if (embeddingList != null) {
```
Yes, because the embeddingList is the DEFAULT_FIELD_CONTENT_VECTOR. It is defined obviously "searchable" but not necessarily "retrievable". It's my case, I don't need to retrieve the vector but only the content and metadata.
So in this line : `List<Double> embeddingList = (List<Double>) searchDocument.get(DEFAULT_FIELD_CONTENT_VECTOR);`  embeddingList is null.

fb33 committed:
chore(IT): apply PR review

make AzureAiSearchEmbeddingStoreIT extends EmbeddingStoreWithFilteringIT

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented on langchain4j-azure-ai-search/src/test/java/dev/langchain4j/rag/content/retriever/azure/search/AzureAiSearchContentRetrieverIT.java:
```diff
@@ -24,7 +24,7 @@
 
 @EnabledIfEnvironmentVariable(named = "AZURE_SEARCH_ENDPOINT", matches = ".+")
-public class AzureAiSearchContentRetrieverIT extends EmbeddingStoreIT {
+public class AzureAiSearchContentRetrieverIT extends EmbeddingStoreWithFilteringIT {
```
Ok

fb33 commented:
Hi,

> I am not sure I understand "searchable/retrievable" thing.
I've put a conditional breakpoint on this line and run all the tests and embeddingList is always not null.
Having this extra if condition is confusing because it is always true (or at least seems to be so from my understanding).

Maybe it's a specificity of Azure Index management. Here is the description of the index fields threw the Azure Portal : 
![image](https://github.com/langchain4j/langchain4j/assets/8268469/cee2cf7b-677e-4b5b-bf61-403ccff338dc)
As you can see, we can decide that a field is searchable (take in account in the search criteria) but is not necessarily retrievable (his value is not returned in the search result)

fb33 commented:
> AFAIU "metadata" and "attributes" are hardcoded inside the AbstractAzureAiSearchEmbeddingStore and cannot be changed, so at this point there is no way to set another names/structure.

This is not critical, but exposing this interface does not seem to have benefits, but it increases the overall complexity.

I understand your point. But I think we can create our own index structure by passing an SearchIndex in the constructor of AzureAiSearchEmbeddingStore, so we are not guaranteed that the default mapping works in every cases.
It's just my thought, I think @jdubois or @agoncal should have the final world :smiley:

fb33 committed:
fix(filter): take review in account

Optimize filter string creation by doing the map once.
https://github.com/langchain4j/langchain4j/pull/1291#discussion_r1662366570

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 commented:
my bad, I push too quickly ! ;-) I'll fix this afternoon

fb33 committed:
fix: use the right filterMapper

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263

fb33 committed:
doc(embedding-store): update comparison table

Update row for Azure AI Search with the feature "Filtering by Metada"

Signed-off-by: François Barbe <f.barbe@lectra.com>

#1263
